lsp-haskell.el

```diff
@@ -246,15 +246,60 @@ and `lsp-haskell-server-args' and `lsp-haskell-server-wrapper-function'."
 ;; https://microsoft.github.io/language-server-protocol/specification#textDocumentItem
 (add-to-list 'lsp-language-id-configuration '(haskell-literate-mode . "haskell"))
 
+(defun lsp-haskell--github-system-type-suffix ()
+  (pcase system-type
+    ('gnu/linux "Linux")
+    ('windows-nt "Windows")
+    ('darwin "macOS")))
+
+(defcustom lsp-haskell-language-server-github-releases-url
+  "https://github.com/haskell/haskell-language-server/releases/latest/download/"
+  "GitHub releases url for haskell-language-server and its binaries. Make sure to include the trailing directory slash."
+  :type 'string
+  :group 'lsp-haskell
+  :package-version '(lsp-mode . "7.1"))
+
+(lsp-dependency
+ 'haskell-language-server-wrapper
+ '(:system "haskell-language-server-wrapper")
+ `(:download :url ,(concat lsp-haskell-language-server-github-releases-url
+			  (format "haskell-language-server-wrapper-%s.gz"
+				  (lsp-haskell--github-system-type-suffix)))
+	     :store-path ,(f-join lsp-server-install-dir "haskell" "haskell-language-server-wrapper")
+	     :decompress :gzip
+	     :set-executable? t))
+
+(defun lsp-haskell--project-ghc-version ()
+  (let* ((path-exe (executable-find "haskell-language-server-wrapper"))
+	 (downloaded-exe (lsp-package-path 'haskell-language-server-wrapper))
+	 (wrapper-exe (if path-exe path-exe downloaded-exe)))
+    (if wrapper-exe
+      (s-trim
+       (shell-command-to-string
+	(format "%s --project-ghc-version 2>/dev/null" wrapper-exe)))
+      nil)))
+
 ;; Register the client itself
 (lsp-register-client
   (make-lsp--client
-    :new-connection (lsp-stdio-connection (lambda () (lsp-haskell--server-command)))
+   :new-connection (lsp-stdio-connection
+		    (lambda ()
+		      (let* ((proj-ghc-ver (lsp-haskell--project-ghc-version))
+			     (server-name (format "haskell-language-server-%s" proj-ghc-ver))
```

Comment:
I think we should support bare `haskell-language-server` (this insists it has the version suffix). People (like me!) who build HLS themselves will just have a binary called `haskell-language-server`.

Possibly this is a use for `lsp-haskell-server-path`: it could be used here if non-nil.